File: java/src/offersix/offersix.java
package offersix;

import java.util.Arrays;

public class offersix {
    public static class ListNode{
        int val;
        ListNode next;
        ListNode(int x) { val = x;}
    }
    /*
    public static int[] reversePrint(ListNode head){
        Stack<Integer> stack = new Stack<>();
        while (head != null) {
            stack.push(head.val);
            head = head.next;
        }
        int []res = new int[stack.size()];
        int i=0;
        while (!stack.empty()) {
            res[i++] = stack.pop();
        }
        return res;
    }
    */
    public static int[] reversePrint(ListNode head){
        int count = 0;
        ListNode tmp = head;
        while (tmp != null) {
            tmp = tmp.next;
            count++;
        }
        int []res = new int[count];
        int j = count - 1;
        while (head != null) {
            res[j] = head.val;
            head = head.next;
            j--;
        }
        return res;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(0);
        ListNode rt = head;
        for(int i = 1; i < 10; i++) {
            ListNode node = new ListNode(i);
            rt.next = node;
            rt = rt.next;
        }
        int [] res = reversePrint(head);
        System.out.println(Arrays.toString(res));
    }

}

File: c++/offersix/offersix.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

/* 栈实现
vector<int> reversePrint(ListNode* head){
    stack<int> Stack;
    ListNode* p = head;
    while (p != NULL) {
        Stack.push(p->val);
        p = p->next;
    }
    vector<int> res;
    while (!Stack.empty()){
        int tmp = Stack.top();
        Stack.pop();
        res.push_back(tmp);        
    }
    return res;
}
*/

vector<int> reversePrint(ListNode* head){
    ListNode* p = head;
    int count = 0;
    while (p != NULL) {
        p = p->next;
        count++;
    }
    vector<int> res(count);
    while (head != NULL) {
        res[--count] = head->val;
        head = head->next;
    }
    return res;
}


int main(){
    ListNode* head = new ListNode(0);
    ListNode* tail = head;
    for (int i=1; i<10; i++) {
        ListNode* node = new ListNode(i);
        tail->next = node;
        tail = tail->next;
    }
    vector<int> res = reversePrint(head);
    for (int i=0; i<res.size(); i++) {
        cout<<res[i]<<" ";
    }
    cout<<endl;
    system("pause");
    return 0;
}